src/levanter/infra/ray_tpu.py

```diff
@@ -0,0 +1,318 @@
+        return TpuPreempted(tpu_info, e)
+    elif isinstance(e, RaySystemError):
+        logger.exception("System error", exc_info=e)
+        return TpuRunError(tpu_info, e)
+    elif isinstance(e, RayTaskError):
+        # node preemptions don't always show up as one of the above errors and can just be a RayTaskError. We have
+        # to try to sniff out the TPU's status.
+        from levanter.infra.tpus import get_current_tpu_is_preempted
+
+        if get_current_tpu_is_preempted():
+            print("Preempted")
+            logger.exception("Preempted", exc_info=e)
+            return TpuPreempted(tpu_info, e)
+
+        logger.exception(f"Task error {e}", exc_info=e)
+        return TpuRunError(tpu_info, e)
+
+    else:
+        logger.exception("Unknown error", exc_info=e)
+        return TpuRunError(tpu_info, e)
+
+
+@dataclass
+class RunOnPodConfig:
+    image_id: str
+    command: list[str] | str
+    tpu_type: str
+    env: dict = dataclasses.field(default_factory=dict)
+    name: str = "levanter"
+    retries: int = 10
+
+
+@draccus.wrap()
+def main(args: RunOnPodConfig):
+    """
+    Run a command on a TPU pod. This is a wrapper around `run_docker_on_pod` that takes a config object as a CLI.
+
+    We use this via infra/launch_on_ray.py to run docker containers on TPUs.
+    """
+    ray.init()
+
+    import shlex
+
+    if isinstance(args.command, str):
+        command = shlex.split(args.command)
+    else:
+        command = args.command
+
+    run_docker_on_pod(
+        args.image_id,
+        command,
+        tpu_type=args.tpu_type,
+        env=args.env,
+        name=args.name,
+    )
+
+
+def _hacky_remove_tpu_lockfile():
+    """
+    This is a hack to remove the lockfile that TPU pods create on the host filesystem.
+
+    libtpu only allows one process to access the TPU at a time, and it uses a lockfile to enforce this.
+    Ordinarily a lockfile would be removed when the process exits, but in the case of Ray, the process is
+    a long-running daemon that doesn't typically exit until the node is shut down. This means that the lockfile
+    persists across Ray tasks. This doesn't apply to our docker-based workloads, but it does apply to other
+    tasks that use JAX directly.
+    """
+    try:
+        os.unlink("/tmp/libtpu_lockfile")
+    except FileNotFoundError:
+        pass
+    except PermissionError:
+        logger.warning("Failed to remove lockfile")
+        try:
+            os.system("sudo rm /tmp/libtpu_lockfile")
+        except Exception:  # noqa
+            pass
+
+
+def _massage_env(env):
+    # Ray pretends it's running in a TTY, which leads to a ton of log spam from tqdm.
+    # Levanter uses tqdm_loggable, which tries to sniff out the TTY, but it doesn't work with Ray.
+    # So we force it
+    if "TERM" not in env:
+        env = {**env, "TERM": "dumb"}
+
+    return env
+
+
+if __name__ == "__main__":
+    main()
+
+    # leaving this here for testing purposes
```

Comment:
Move this into a test file?

Reply:
i mean, it's more of a scratch thing than a unit test